OpenLineage + Airflow guide (#349)
astronomer/airflow-guides: guides/airflow-openlineage.md

It follows that data lineage is a natural integration with Apache Airflow. Airflow is often used as a one-stop-shop orchestrator for an organization’s data pipelines, and is an ideal place for integrating data lineage to understand the movement and interactions of your data.

In this guide, we’ll define data lineage, review the OpenLineage standard and core lineage concepts, describe why you would want data lineage with Airflow, and show an example local integration of Airflow and OpenLineage for tracking data movement in Postgres.

> Note: This guide focuses on using OpenLineage with Airflow 2. Some topics may differs lightly if you are using it with earlier versions of Airflow.

## What is Data Lineage

> According to [Datakin](https://datakin.com/), the creators of the OpenLineage standard and leaders in the data lineage space, data lineage is “the complicated set of relationships between datasets”.  The concept of data lineage encompasses:

Author: Great question. @rossturk do you have a sense of what our language should be around Datakin now post-announcement? Do we still refer to it as a separate entity, will our websites be integrated, etc.?

> According to [Datakin](https://datakin.com/), the creators of the OpenLineage standard and leaders in the data lineage space, data lineage is “the complicated set of relationships between datasets”.  The concept of data lineage encompasses:

Author: Sounds good, will update to that language. When I originally wrote it I mostly wanted to make sure I wasn't plagiarizing directly from Datakin's stuff, but it's a little easier now that we're publishing stuff together :)

- **Lineage metadata** that describes your datasets (e.g. a table in Snowflake) and jobs (e.g. tasks in your DAG).
- **The lineage graph** that shows a visual representation of your jobs and datasets and how they are connected.

If you want to read more on the concept of data lineage and why it’s important, check out this [Datakin blog post](https://datakin.com/what-is-data-lineage/).

Visually, your data lineage graph might look something like this:

![Lineage Graph](https://assets2.astronomer.io/main/guides/airflow-openlineage/example_lineage_graph.png)

> If you are using data lineage, you will likely have a lineage tool that collects lineage metadata, as well as a front end for visualizing the lineage graph. There are paid tools (including Datakin) that provide these services, but in this guide we will focus on the open source options that can be integrated with Airflow: namely OpenLineage (the lineage tool) and [Marquez](https://marquezproject.github.io/marquez/) (the lineage front end).

Author: I originally had a section in here for lineage on Astro, but took it out for now since that isn't available just yet. I don't want to hold up publishing the guide since we wanted it for marketing/PR purposes this week. But as soon as we have something in Astro I plan to update to reference that.

- Quickly find the root cause of task failures by identifying issues in upstream datasets (e.g. if an upstream job outside of Airflow failed to populate a key dataset).
- Easily see the blast radius of any job failures or changes to data by visualizing the relationship between jobs and datasets.
- Identify where key data is used in jobs across an organization.

These capabilities translate into real world benefits by:

- Making recovery from complex failures quicker. The faster you can identify the problem and the blast radius, the easier it is to solve and prevent any erroneous decisions being made from bad data.
- Making it easier for teams to work together across an organization. Visualizing the full scope of where a dataset is used reduces “sleuthing” time.
- Helping ensure compliance with data regulations by fully understanding where data is used.

## Example Integration

In this example, we’ll show how to run OpenLineage with Airflow locally using Marquez as a lineage front end. We’ll then show how to use and interpret the lineage data generated by a simple DAG that processes data in Postgres.

### Running Marquez and Airflow Locally

> For this example, we’ll run Airflow with OpenLineage and Marquez locally following the steps below. You will need Docker and the [Astro CLI](https://docs.astronomer.io/cloud/install-cli) installed before starting.

Reviewer: ```suggestion
For this example, we’ll run Airflow with OpenLineage and Marquez locally. You will need to install Docker and the [Astro CLI](https://docs.astronomer.io/cloud/install-cli) before starting.
```

> 1. Run Marquez locally using their [quickstart instructions](https://github.com/MarquezProject/marquez#quickstart): clone the repo, then run `./docker/up.sh` from the root directory.

Reviewer: ```suggestion
1. Run Marquez locally using the quickstart in the [Marquez README](https://github.com/MarquezProject/marquez#quickstart).
```

> 2. Start an astro project using the CLI by creating a new directory and running `astrocloud dev init`.

Reviewer: ```suggestion
2. Start an Astro project using the CLI by creating a new directory and running `astrocloud dev init`.
```

> 3. Add `openlineage-airflow` to your `requirements.txt` file. Note if you are using Astronomer Runtime 4.0.9 or greater, this package is already included.

Reviewer: ```suggestion
3. Add `openlineage-airflow` to your `requirements.txt` file. Note if you are using Astro Runtime 4.0.9 or greater, this package is already included.
```

4. Add the environment variables below to your `.env` file. These will allow Airflow to connect with the OpenLineage API and send your lineage data to Marquez.
    
    ```bash
    OPENLINEAGE_URL=http://host.docker.internal:5000
    OPENLINEAGE_NAMESPACE=example
    AIRFLOW__LINEAGE__BACKEND=openlineage.lineage_backend.OpenLineageBackend
    ```
    

>     Port 5000 is used by Marquez by default when you run it using Docker in Step 1. If you are not using Marquez and are using a different OpenLineage front end, or are running Marquez remotely, you can modify the `OPENLINEAGE_URL` as needed.

Reviewer: ```suggestion
    By default, Marquez uses port 5000 when you run it using Docker. If you are using a different OpenLineage front end instead of Marquez, or you are running Marquez remotely, you can modify the `OPENLINEAGE_URL` as needed.
```

![Marquez Jobs](https://assets2.astronomer.io/main/guides/airflow-openlineage/marquez_jobs.png)

Then, if we click on one of the jobs from our DAGs, we see the full lineage graph.

![Marquez Graph](https://assets2.astronomer.io/main/guides/airflow-openlineage/marquez_lineage_graph.png)

The lineage graph shows:

- Two origin datasets that are used to populate the combined data table
- The two jobs (tasks) from our DAGs that result in new datasets: `combine` and `reporting`
- Two new datasets that are created by those jobs

The lineage graph shows us how these two DAGs are connected and how data flows through the entire pipeline, giving us insight we wouldn't have if we were to view these DAGs in the Airflow UI alone.

## Limitations

OpenLineage is rapidly evolving, and new functionality and integrations are being added all the time. At the time of writing, the following are limitations when using OpenLineage with Airflow:

- The OpenLineage integration with Airflow 2 is currently experimental, and only reports lineage data for *successful* task runs.

> - Only the following operators have bundled extractors (needed to collect lineage data out of the box):

Reviewer: Maybe we should add another point here like "To get lineage from an external system connected to Airflow, such as Spark, you need to complete an integration with that system in addition to Airflow".

Author: Good call, added that bullet with some relevant links

It follows that data lineage is a natural integration with Apache Airflow. Airflow is often used as a one-stop-shop orchestrator for an organization’s data pipelines, and is an ideal place for integrating data lineage to understand the movement and interactions of your data.

In this guide, we’ll define data lineage, review the OpenLineage standard and core lineage concepts, describe why you would want data lineage with Airflow, and show an example local integration of Airflow and OpenLineage for tracking data movement in Postgres.

> Note: This guide focuses on using OpenLineage with Airflow 2. Some topics may differs lightly if you are using it with earlier versions of Airflow.

## What is Data Lineage

According to [Datakin](https://datakin.com/), the creators of the OpenLineage standard and leaders in the data lineage space, data lineage is “the complicated set of relationships between datasets”.  The concept of data lineage encompasses:

- **Lineage metadata** that describes your datasets (e.g. a table in Snowflake) and jobs (e.g. tasks in your DAG).

> - **A lineage frontend** that allows you to view and interact with your lineage metadata, including a graph that visualizes your jobs and datasets and shows how they are connected.

Author: Good call, added that bullet

    
    By default, Marquez uses port 5000 when you run it using Docker. If you are using a different OpenLineage front end instead of Marquez, or you are running Marquez remotely, you can modify the `OPENLINEAGE_URL` as needed.
    
5. Modify your `config.yaml` in the `.astrocloud/` directory to choose a different port for Postgres. Marquez also uses Postgres, so you will need to have Airflow use a different port than the default 5432, which will already be allocated.
    
    ```yaml
    project:
      name: openlineage
    postgres:
      port: 5435
    ```
    
6. Run Airflow locally using `astrocloud dev start`.
7. Confirm Airflow is running by going to `http://localhost:8080`, and Marquez is running by going to `http://localhost:3000`.

### Generating and Viewing Lineage Data

To show the lineage data that can result from Airflow DAG runs, we'll use an example of two DAGs that process data in Postgres. The first DAG creates and populates a table (`animal_adoptions_combined`) with data aggregated from two source tables (`adoption_center_1` and `adoption_center_2`):

```python
from airflow import DAG
from airflow.providers.postgres.operators.postgres import PostgresOperator

> import random

Author: Nope, a leftover from playing with Willy's example :)
